Cython/Utility/ObjectHandling.c

```diff
@@ -2000,15 +2019,53 @@ static int __Pyx_TryUnpackUnboundCMethod(__Pyx_CachedCFunction* target) {
     target->method = method;
 #if CYTHON_COMPILING_IN_CPYTHON
     #if PY_MAJOR_VERSION >= 3
-    // method descriptor type isn't exported in Py2.x, cannot easily check the type there
     if (likely(__Pyx_TypeCheck(method, &PyMethodDescr_Type)))
+    #else
+    // method descriptor type isn't exported in Py2.x, cannot easily check the type there.
+    // Therefore, reverse the check to the most likely alternative
+    // (which is returned for class methods)
+    if (likely(!PyCFunction_Check(method)))
     #endif
     {
         PyMethodDescrObject *descr = (PyMethodDescrObject*) method;
         target->func = descr->d_method->ml_meth;
         target->flag = descr->d_method->ml_flags & ~(METH_CLASS | METH_STATIC | METH_COEXIST | METH_STACKLESS);
-    }
+    } else
+#endif
+    // bound classmethods need special treatment
+#if defined(CYTHON_COMPILING_IN_PYPY)
+    // In these functions are regular methods, so just do
+    // the self check
+#elif PY_VERSION_HEX >= 0x03090000
+    if (PyCFunction_CheckExact(method))
+#else
+    if (PyCFunction_Check(method))
 #endif
+    {
+        PyObject *self;
+        int self_found;
+#if CYTHON_COMPILING_IN_LIMITED_API || CYTHON_COMPILING_IN_PYPY
+        self = PyObject_GetAttrString(method, "__self__");
+        if (!self) {
+            PyErr_Clear();
+        }
+#else
+        self = PyCFunction_GET_SELF(method);
+#endif
+        self_found = (self && self != Py_None);
+#if CYTHON_COMPILING_IN_LIMITED_API || CYTHON_COMPILING_IN_PYPY
+        Py_XDECREF(self);
+#endif
+        if (self_found) {
+            PyObject *unbound_method = PyCFunction_New(&__Pyx_UnboundCMethod_Def, method);
+            if (unlikely(!unbound_method)) {
+                Py_DECREF(method);
+                return -1;
+            }
+            Py_INCREF(method);  // PyCFunction will Py_XDECREF method on destruction
```

Comment:
If I think about it it seems wrong, `__Pyx_PyObject_GetAttrStr` should have already incremented the refcount so we should be fine then?

Reply:
`PyCFunction_New` increfs self internally, so it `XDECREF`s its own reference. We want to `Py_DECREF` method on both the error and non-error path after `PyCFunction_New` I think

Reply:
Oh yeah, I missed that `Py_XNewRef` for self.
I think you are correct, previosly `method` was not cleaned up on a non-error path

This maked me think if we can have some test on how much memory is still occupied before python terminates, it would be nice to catch those kinds of errors, this is likely to be a systematic problem

Also I wonder if there is a valgrind test, it still seems like a good idea to have it at least for one platform

Reply:
I don't think Python cleans up after itself perfectly on shutdown anyway so I believe Valgrind is quite hard to interpret.

We have refnanny for references in generated code, but not utility code.

Reply:
I tried to experiment with [cpython/Misc/Readme.valgrind](https://github.com/python/cpython/blob/main/Misc/README.valgrind) and [cpython/Misc/valgrind-python.supp](https://github.com/python/cpython/blob/main/Misc/valgrind-python.supp), but it's not very fruitful 
I tried to generate a suppression from the scratch, and I'm getting mixed results

Reply:
Actually this should be a separate issue

Reply:
> We have refnanny for references in generated code, but not utility code.

Note that you can use it also in utility code, if you put in the required declarations and use the refnanny incref/decref/gotref/giveref macros. We just don't normally do it because it's cumbersome to copy all the boiler plate code together and write refnanny code by hand. If it helps here, it might be worth using. (My guess is that there are quite a number of utility code functions that might benefit from this, especially in `ObjectHandling.c` and `Builtins.c`, which deal a lot with object references.

Reply:
resolving in favor of #5224